### PR TITLE
[LLDB] Ignore actual-needed artificial members in DWARFASTParserClang::ParseSingleMember

### LINKED ISSUE #69309: [lldb] No pretty print for the coroutine frame type.
The lldb doesn't seem to know about the coroutine frame type (the structure type [`<function_name>.coro_frame_ty`](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Coroutines/CoroFrame.cpp#L1134) emitting in the clang middle-end). See the poor print results below.

```
$ ./bin/clang++ -std=c++20 -g /tmp/t.cpp -o debugging-example 
$ lldb debugging-example 
...
(lldb) r
...
Process 1790445 stopped
* thread #1, name = 'debugging-examp', stop reason = signal SIGTRAP
    frame #0: 0x00005555555555fc debugging-example`coro_task(v=1) at t.cpp:46:3
   43    task coro_task(int v) {
   44     int a = v;
   45     __builtin_debugtrap();
-> 46     co_return;
   47   }
   48  
   49   int main() {
(lldb) p __coro_frame
(_Z9coro_taski.coro_frame_ty)  {}
```

The gdb gives better results on printing the `__coro_frame`:

```
(gdb) p __coro_frame
$1 = {__resume_fn = 0x555555555bf0 <detail::chain_fn<0>()>, __destroy_fn = 0x555555555d90 <detail::chain_fn<0>()>, __promise = {continuation = {_M_fr_ptr = 0x555555574440}, result = 0}, 
  struct_task__promise_type__FinalSuspend_0 = {struct_std____n4861__coroutine_handle_0 = {PointerType = 0x0}}, __coro_index = 0 '\000', struct_std____n4861__suspend_always_1 = {__int_8 = 0 '\000'}}
```

```
// tmp/t.cpp
#include <coroutine>

struct task{
  struct promise_type {
    task get_return_object() { return std::coroutine_handle<promise_type>::from_promise(*this); }
    std::suspend_always initial_suspend() { return {}; }
    std::suspend_always final_suspend() noexcept { return {}; }
    void return_void() noexcept {}
    void unhandled_exception() noexcept {}

    int count = 0;
  };

  void resume() noexcept {
    handle.resume();
  }

  task(std::coroutine_handle<promise_type> hdl) : handle(hdl) {}
  ~task() {
    if (handle)
      handle.destroy();
  }

  std::coroutine_handle<> handle;
};

 task coro_task(int v) {
  int a = v;
  __builtin_debugtrap();
  co_return;
}

int main() {
  auto k = coro_task(1);
  k.resume();
  return 0;
}
```

### PR DESCRIPTION
**Pull request by hokein**
Address the FIXME, this will allow lldb to print all fields of the generated coroutine frame structure.

Fixes #69309.

### CODE REVIEW

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp [hunk 1]
```diff
@@ -41,7 +41,10 @@ CPPLanguageRuntime::CPPLanguageRuntime(Process *process)
     : LanguageRuntime(process) {}
 
 bool CPPLanguageRuntime::IsAllowedRuntimeValue(ConstString name) {
-  return name == g_this;
+  // FIXME: use a list when the list grows more.
+  return name == g_this ||
+         name == ConstString("__promise") ||
+         name == ConstString("__coro_frame");
```
Comment (vogelsgesang) [inline lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:47]: I am not sure if I would expose the `__coro_frame` variable. Maybe I would only expose the `__promise` variable by default. The `__coro_frame` is usually out-of-sync and does not hold the up-to-date coroutine state while the coroutine is in flight

**Review by vogelsgesang — COMMENTED** [review]
I can't provide feedback on the lldb integration aspects, as I don't know the lldb code base well enough

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp (hunk 1 above)
Comment (Michael137) [reply lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:47]: Any chance we can get the `Type` from the `Variable` in `CommandObjectFrameVariable` and can re-use the `IsCoroutineFrameType` check?

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 2]
```diff
@@ -3061,7 +3061,8 @@ void DWARFASTParserClang::ParseSingleMember(
   // artificial member with (unnamed bitfield) padding.
   // FIXME: This check should verify that this is indeed an artificial member
   // we are supposed to ignore.
-  if (attrs.is_artificial) {
+  if (attrs.is_artificial &&
+      !TypeSystemClang::IsCoroutineFrameType(class_clang_type)) {
```
Comment (Michael137) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:3065]: I do still wonder if we should just make a `ShouldIgnoreArtificialField` API that returns `true` for known fields we want to skip. The only ones that comes to mind are artificial variables that start with `_vptr$`

Not sure if there are some Objective-C artificial vars we want to skip. Maybe @jimingham or @adrian-prantl would know

Comment (hokein) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp]: I changed to `ShouldIgnoreArtificialField` per suggest (unite tests are passed). It looks like the vtable pointer is the only case for C/C++.

**Review by hokein — COMMENTED** [review]
Thanks for the review.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 3]
```diff
@@ -3059,9 +3059,8 @@ void DWARFASTParserClang::ParseSingleMember(
   // This needs to be done after updating FieldInfo which keeps track of where
   // field start/end so we don't later try to fill the space of this
   // artificial member with (unnamed bitfield) padding.
-  // FIXME: This check should verify that this is indeed an artificial member
-  // we are supposed to ignore.
-  if (attrs.is_artificial) {
+  if (attrs.is_artificial &&
+      TypeSystemClang::ShouldIgnoreArtificialField(attrs.name)) {
```
Comment (Michael137) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:3063]: We'll need to double check how clang's layout builder is affected by not ignoring these artificial fields

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp (hunk 1 above)
Comment (hokein) [reply lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp]: I removed this change from this pull request now (will sent as a followup). Let's focus on addressing the ignore-artificial-fields issue in `DWARFASTParserClang.cpp`.

File: lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp [hunk 4]
```diff
@@ -771,6 +771,10 @@ TypeSystemClang *TypeSystemClang::GetASTContext(clang::ASTContext *ast) {
   return clang_ast;
 }
 
+bool TypeSystemClang::ShouldIgnoreArtificialField(llvm::StringRef Name) {
+  return Name.starts_with("_vptr$");
```
[the remark this answers is not in the document]
Comment (hokein) [reply lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp]: Thanks, good catch. Handled the gdb case and added a test case (I thought the Clang name in the filename and path name indicates that these part of code only cares about clang-generated code, but it looks like we use Clang type system to parse the DWARF debug info).

Do we care about the msvc-built program as well? I think? probably not, as we use a different debug info format `PDB` on Windows. And this class is only used in parsing DWARF.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp (hunk 3 above)
Comment (hokein) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:3063]: This function is only used to parse `DW_TAG_MEMBER`. I took a look on all related usages of `createMemberType` with `llvm::DINode::FlagArtificial` in the llvm source repo:
- vptr pointer: https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGDebugInfo.cpp#L2426
- coroutines:
   - https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Coroutines/CoroFrame.cpp#L1058  
   - https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Coroutines/CoroFrame.cpp#L1234

so two cases are virtual pointer, and coroutines frames. And to my best knowledge, not ignoring coroutine artificial fields should be fine, the coroutine frame is just a regular plain struct .

**Comment by hokein** [comment]
Thanks for the comment, I updated the patch, this is ready for another round of review.

File: lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Michael137) [reply lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp]: `TypeSystemClang` should be agnostic to the debug-info format. The various debug-info->AST parsers (e.g., `DWARFASTParserClang`/`PDBASTParser`/`npdb::PdbAstBuilder`, the latter two being the MSVC debug-info parsers) use `TypeSystemClang` after they've parsed debug-info to construct clang AST nodes. But I'm not sure how the PDB parsers handle the vptr case and artificial variables. From a brief glance it doesn't look like it cares about artificial member variables at all.

[the remark this answers is not in the document]
Comment (Michael137) [reply lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp]: Are you still planning on using this API from the `CPPLanguageRuntime`? If not, you could just put it inside of `DWARFASTParserClang.cpp`.

[the remark this answers is not in the document]
Comment (hokein) [reply lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp]: > TypeSystemClang should be agnostic to the debug-info format. The various debug-info->AST parsers (e.g., DWARFASTParserClang/PDBASTParser/npdb::PdbAstBuilder, the latter two being the MSVC debug-info parsers) use TypeSystemClang after they've parsed debug-info to construct clang AST nodes.

Thanks for the clarification. 

That makes sense. I don't have any plan to use this API. It is better to move DWARFASTParserClang.cpp (better layering).

File: lldb/test/Shell/SymbolFile/DWARF/ignored_artificial_fields.test [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# UNSUPPORTED: system-darwin, system-windows
```
Comment (Michael137) [inline lldb/test/Shell/SymbolFile/DWARF/ignored_artificial_fields.test:1]: Why unsupported on Darwin? Pretty sure it should just work

Comment (Michael137) [reply lldb/test/Shell/SymbolFile/DWARF/ignored_artificial_fields.test:1]: I think we can just remove these and if the test does end up failing we'll just add the requirements back after the fact.

Comment (Michael137) [reply lldb/test/Shell/SymbolFile/DWARF/ignored_artificial_fields.test:1]: Oh you're using gcc for this. Nvm

**Review by Michael137 — APPROVED** [review]
LGTM

Are you still going to make the change to `CommandObjectFrame` such that `frame var` doesn't hide the artificial fields?

**Comment by hokein** [comment]
Thank you very much for the review!

> Are you still going to make the change to CommandObjectFrame such that frame var doesn't hide the artificial fields?

Yes, that's my next step.

**Comment by DavidSpickett** [comment]
Hi, this is failing on Linaro's Linux LLDB bots: https://lab.llvm.org/buildbot/#/builders/96/builds/48277

Looks like the `-m64` (or `-m32`) is being passed to an AArch64 (or ARM) g++, and those are X86 only flags.

**Comment by DavidSpickett** [comment]
I think I have a fix for this, will commit it soon.

**Comment by hokein** [comment]
> Hi, this is failing on Linaro's Linux LLDB bots: https://lab.llvm.org/buildbot/#/builders/96/builds/48277

> Looks like the -m64 (or -m32) is being passed to an AArch64 (or ARM) g++, and those are X86 only flags.

Sorry for the breakage. I removed this test temporarily in 343eb4b4253fea31767f6a98e1cf77a7d69c856a to make the builtbots happy again while doing the investigation.

**Comment by DavidSpickett** [comment]
I just pushed a fix for the build script, I'll put the test case back once I know the script change is good.

**Comment by hokein** [comment]
Thank you very much for the fix!

**Comment by DavidSpickett** [comment]
All fixed now.

**Comment by Caslyn** [comment]
Hi @hokein - ignored_artificial_fields.test breaks if gcc is unavailable. For ex we purposely exclude gcc on our downstream bots so as not to implicitly depend on it. Maybe all that is needed is adding `# REQUIRES: gcc`.

Would you be able to revert the test and/or patch it to only run when gcc is present?

**Comment by hokein** [comment]
@Caslyn , I added the `# REQUIRES: gcc` to the test in https://github.com/llvm/llvm-project/commit/81a76902ae0fc08138e37212239c5c704eec2f26

**Pull request merged**